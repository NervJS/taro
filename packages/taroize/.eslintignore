/lib
/index.js
s.js
t.js