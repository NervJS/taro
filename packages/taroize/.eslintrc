{
  "rules": {
    "arrow-spacing": "error",
    "key-spacing": "error",
    "keyword-spacing": "error",
    "space-before-blocks": "error",
    "space-infix-ops": "error"
  }
}