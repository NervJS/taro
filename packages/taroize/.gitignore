lib
t.js
*.wxml
s.js