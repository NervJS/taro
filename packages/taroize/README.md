# `@tarojs/taroize`

暴露给 `@tarojs/cli/convert` 的原生小程序转 Taro 转换器。

## 核心 API

### parse(options)

解析小程序源文件为 React/Vue 组件、

#### options

```typescript
interface Option {
  json?: string, // 小程序配置文件
  script?: string, // 小程序的 JavaScript 文件
  wxml?: string, // wxml 文件
  path: string, // 当前页面路径
  rootPath: string // 根目录路径
}
```
