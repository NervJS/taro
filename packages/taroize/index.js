module.exports = require('./lib/src/index').parse
