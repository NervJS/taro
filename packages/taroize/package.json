{
  "name": "@tarojs/taroize",
  "version": "1.2.7",
  "description": "转换原生微信小程序代码为 Taro 代码",
  "main": "index.js",
  "files": [
    "index.js",
    "lib"
  ],
  "scripts": {
    "test": "jest",
    "build": "tsc",
    "dev": "tsc -w --pretty"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/NervJS/taro/tree/master/packages/taroize"
  },
  "author": "yuche",
  "license": "MIT",
  "dependencies": {
    "@babel/code-frame": "^7.0.0",
    "babel-core": "^6.26.3",
    "babel-generator": "^6.26.1",
    "babel-template": "^6.26.0",
    "babel-traverse": "^6.26.0",
    "babel-types": "^6.26.0",
    "babylon": "^6.18.0",
    "himalaya-wxml": "^1.1.0",
    "lodash": "^4.17.5",
    "typescript": "^3.0.1"
  },
  "devDependencies": {
    "@types/babel-generator": "^6.25.2",
    "@types/babel-traverse": "^6.25.4",
    "@types/babel-types": "^7.0.4",
    "@types/jest": "^22.2.3",
    "@types/lodash": "^4.14.105",
    "@types/node": "^9.6.2",
    "better-babel-generator": "^6.26.1",
    "jest": "^23.0.1",
    "jest-cli": "^22.1.4",
    "prettier": "^1.14.2",
    "ts-jest": "^22.4.6",
    "tslint": "^5.10.0",
    "tslint-config-prettier": "^1.10.0",
    "tslint-config-standard": "^7.0.0"
  },
  "publishConfig": {
    "access": "public"
  }
}
