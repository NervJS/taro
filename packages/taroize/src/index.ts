import * as t from 'babel-types'
import { parseWXML } from './wxml'
import { parseScript } from './script'
import { parseJSON } from './json'
import { errors } from './global'

interface Option {
  json?: string,
  script?: string,
  wxml?: string,
  path: string
}

export function parse (option: Option) {
  const { wxml, wxses, imports, refIds } = parseWXML(option.path, option.wxml)
  const json = parseJSON(option.json)
  const ast = parseScript(option.script, wxml as t.Expression, json, wxses, refIds)
  return {
    ast,
    imports,
    errors
  }
}
