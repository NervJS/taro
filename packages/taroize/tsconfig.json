{
  "compilerOptions": {
    "target": "es2017",
    "module": "commonjs",
    "removeComments": false,
    "preserveConstEnums": true,
    "moduleResolution": "node",
    "experimentalDecorators": true,
    "noImplicitAny": false,
    "allowSyntheticDefaultImports": true,
    "outDir": "lib",
    "noUnusedLocals": true,
    "noUnusedParameters": true,
    "strictNullChecks": true,
    "sourceMap": true,
    "baseUrl": ".",
    "rootDir": "."
  },
  "include": ["src"],
  "exclude": [
    "__tests__",
    "node_modules",
    "dist",
    "tests",
    "jest",
    "lib",
    "**/*.test.ts",
    "**/*.spec.ts"
  ],
  "compileOnSave": false
}
