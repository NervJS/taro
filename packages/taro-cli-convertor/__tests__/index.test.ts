import * as taroize from '@tarojs/taroize'

import Convertor from '../src/index'

describe('parseAst', () => {
  test('当使用e.target.dataset时引入工具函数 getTarget', () => {
    const entryJSON = { 'pages': ['pages/index/index'] }
    /**
     * json：index.json的内容
     * path：index的根目录（文件路径)
     * rootPath：小程序的根目录（文件路径）
     * script：index.js的内容
     * scriptPath：index.js的绝对路径
     * wxml：index.html的内容
     */
    const param = {
      json:'{}',
      path:'',
      rootPath:'',
      script:`
        app.createPage({
          data:{
            tagInfo:{
              id:123456,
              data:'茅台',
            },
            msg:'',
            tagName:{}  
          },
          getMsg(e){
            const detail = e.currentTarget
            //变量赋值
            const tagName_ = e.currentTarget.dataset.tagName
            const tagData_ = e.currentTarget.dataset.tagData
            //结构赋值
            const { tagName } = e?.target?.dataset
            const { tagData } = e?.target?.dataset
          },
          getMsg02(e){
            const detail = e.currentTarget
            const tagname_ = e.currentTarget.dataset.tagname
            const tagdata_ = e.currentTarget.dataset.tagdata
            const { tagname } = detail?.dataset
            const { tagdata } = detail?.dataset
          },
        })
      `,
      scriptPath:'',
      wxml:`
        <view>测试data-xxx-xxx写法</view>
        <button data-tag-name="WX1314" data-tag-data="{{ tagInfo }}" bindtap="getMsg">获取</button>
        <view>测试data-xxxXxxx 驼峰写法</view>
        <button data-tagName="WX1314" data-tagData="{{ tagInfo }}" bindtap="getMsg02">获取</button>
        `
    }
    // new Convertot后会直接执行 init()，为确保 init() 在测试中通过采用 spyOn 去模拟
    jest.spyOn(Convertor.prototype, 'getApp').mockImplementation(() => {
      Convertor.prototype.entryJSON = entryJSON
    })
    jest.spyOn(Convertor.prototype, 'getPages').mockImplementation(() => {
      Convertor.prototype.pages = new Set(entryJSON.pages)
    })
    const convert = new Convertor('', false)
    convert.pages = Convertor.prototype.pages
    const taroizeResult = taroize({
      ...param,
      framework:'react'
    })
    /**
     * sourceFilePath：需要转换的文件路径
     * outputFilePath：转换输出路径
     * importStylePath：style的文件路径
     */
    const { ast } = convert.parseAst({
      ast:taroizeResult.ast,
      sourceFilePath:'',
      outputFilePath:'',
      importStylePath:'',
      depComponents: new Set(),
      imports:[]
    })
    expect(ast).toMatchSnapshot()
  })
})