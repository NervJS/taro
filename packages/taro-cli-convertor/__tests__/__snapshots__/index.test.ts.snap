// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseAst require引用为空 1`] = `
import withWeapp, { cacheOptions } from "@tarojs/with-weapp";
import { Block } from "@tarojs/components";
import React from "react";
import Taro from "@tarojs/taro";
const aa = require();
cacheOptions.setOptionsToCache({});
@withWeapp(cacheOptions.getOptionsFromCache())
class _C extends React.Component {
  render() {
    return null;
  }
}
export default _C;
`;

exports[`parseAst 为data数据添加可选链操作符 1`] = `
import withWeapp, { cacheOptions } from "@tarojs/with-weapp";
import { Block, View } from "@tarojs/components";
import React from "react";
import Taro from "@tarojs/taro";
cacheOptions.setOptionsToCache({
  data: {
    testString: '',
    testArray: [],
    testUnArray: undefined
  }
});
@withWeapp(cacheOptions.getOptionsFromCache())
class _C extends React.Component {
  render() {
    const {
      testString,
      testArray,
      testUnArray
    } = this.data;
    return <Block><View>{testString?.trim()}</View>{testArray?.[0]?.children?.map((item, index) => {
        return <View>{item}</View>;
      })}{testUnArray?.[1]?.children?.map((item, index) => {
        return <View>{item}</View>;
      })}</Block>;
  }
}
export default _C;
`;

exports[`parseAst 为setData数据添加可选链操作符 1`] = `
import withWeapp, { cacheOptions } from "@tarojs/with-weapp";
import { Block, View } from "@tarojs/components";
import React from "react";
import Taro from "@tarojs/taro";
cacheOptions.setOptionsToCache({
  onLoad() {
    let list = [];
    let dcopy = [];
    list.push({
      children: [0, 1, 2, 3, 4, 5]
    });
    list.push({
      children: [0, 1, 2, 3, 4]
    });
    this.setData({
      list,
      dcopy: [0, 1, 2, 3, 4, 5]
    });
  }
});
@withWeapp(cacheOptions.getOptionsFromCache())
class _C extends React.Component {
  render() {
    const {
      list,
      dcopy
    } = this.data;
    return <Block>{list?.[0]?.children?.map((item, index) => {
        return <View>{item}</View>;
      })}{dcopy?.map((item, index) => {
        return <View>{item}</View>;
      })}</Block>;
  }
}
export default _C;
`;

exports[`parseAst 使用 resolveAlias 配置项用来自定义模块路径的映射规则 1`] = `
import withWeapp, { cacheOptions } from "@tarojs/with-weapp";
import { Block } from "@tarojs/components";
import React from "react";
import Taro from "@tarojs/taro";
const {
  formatTime
} = require("../tools/util.js");
import { mesg } from "./utils.js";
const {
  test
} = require("../../components/navigation-bar/test.js");
cacheOptions.setOptionsToCache({});
@withWeapp(cacheOptions.getOptionsFromCache())
class _C extends React.Component {
  render() {
    return null;
  }
}
export default _C;
`;

exports[`parseAst 处理js文件中以/开头的绝对路径 1`] = `
import withWeapp, { cacheOptions } from "@tarojs/with-weapp";
import { Block } from "@tarojs/components";
import React from "react";
import Taro from "@tarojs/taro";
const {
  add
} = require("./add.js");
cacheOptions.setOptionsToCache({});
@withWeapp(cacheOptions.getOptionsFromCache())
class _C extends React.Component {
  render() {
    return null;
  }
}
export default _C;
`;

exports[`parseAst 处理js文件中非正常路径，比如 a/b 1`] = `
import withWeapp, { cacheOptions } from "@tarojs/with-weapp";
import { Block } from "@tarojs/components";
import React from "react";
import Taro from "@tarojs/taro";
const {
  add
} = require("./add.js");
cacheOptions.setOptionsToCache({});
@withWeapp(cacheOptions.getOptionsFromCache())
class _C extends React.Component {
  render() {
    return null;
  }
}
export default _C;
`;

exports[`parseAst 对图片路径的转换处理 1`] = `
import withWeapp, { cacheOptions } from "@tarojs/with-weapp";
import { Block, Image } from "@tarojs/components";
import React from "react";
import Taro from "@tarojs/taro";
cacheOptions.setOptionsToCache({});
@withWeapp(cacheOptions.getOptionsFromCache())
class _C extends React.Component {
  render() {
    return <Image src={require("../../images/tutu.jpg")} mode="scaleToFill"></Image>;
  }
}
export default _C;
`;

exports[`parseAst 当使用e.target.dataset时引入工具函数 getTarget 1`] = `
import withWeapp, { cacheOptions, getTarget } from "@tarojs/with-weapp";
import { Block, View, Button } from "@tarojs/components";
import React from "react";
import Taro from "@tarojs/taro";
app.createPage({
  data: {
    tagInfo: {
      id: 123456,
      data: '茅台'
    },
    msg: '',
    tagName: {}
  },
  getMsg(e) {
    const detail = e.currentTarget;
    //变量赋值
    const tagName_ = getTarget(e.currentTarget, Taro).dataset.tagName;
    const tagData_ = getTarget(e.currentTarget, Taro).dataset.tagData;
    //结构赋值
    const {
      tagName
    } = getTarget(e?.target, Taro)?.dataset;
    const {
      tagData
    } = getTarget(e?.target, Taro)?.dataset;
  },
  getMsg02(e) {
    const detail = e.currentTarget;
    const tagname_ = getTarget(e.currentTarget, Taro).dataset.tagname;
    const tagdata_ = getTarget(e.currentTarget, Taro).dataset.tagdata;
    const {
      tagname
    } = getTarget(detail, Taro)?.dataset;
    const {
      tagdata
    } = getTarget(detail, Taro)?.dataset;
  }
});
cacheOptions.setOptionsToCache({});
@withWeapp(cacheOptions.getOptionsFromCache())
class _C extends React.Component {
  render() {
    const {
      tagInfo
    } = this.data;
    return <Block><View>测试data-xxx-xxx写法</View><Button data-tag-name="WX1314" data-tag-data={tagInfo} onClick={this.getMsg}>获取</Button><View>测试data-xxxXxxx 驼峰写法</View><Button data-tagname="WX1314" data-tagdata={tagInfo} onClick={this.getMsg02}>获取</Button></Block>;
  }
}
export default _C;
`;

exports[`parseAst 模板中使用function 1`] = `
import withWeapp, { cacheOptions } from "@tarojs/with-weapp";
import { Block, Button } from "@tarojs/components";
import React from "react";
import Taro from "@tarojs/taro";
import ItemTmpl from '../../../imports/ItemTmpl.js';
cacheOptions.setOptionsToCache({
  onClick() {
    console.log('click');
  }
});
@withWeapp(cacheOptions.getOptionsFromCache())
class _C extends React.Component {
  render() {
    return <ItemTmpl data={{
      text: 'forbar'
    }} onClick={this.onClick}></ItemTmpl>;
  }
}
export default _C;
`;

exports[`parseAst 模板按需导入 1`] = `
import withWeapp, { cacheOptions } from "@tarojs/with-weapp";
import { Block } from "@tarojs/components";
import React from "react";
import Taro from "@tarojs/taro";
import TemplateATmpl from 'taroConvert/src/imports/TemplateATmpl.js';
cacheOptions.setOptionsToCache({});
@withWeapp(cacheOptions.getOptionsFromCache())
class _C extends React.Component {
  render() {
    return <TemplateATmpl></TemplateATmpl>;
  }
}
export default _C;
`;

exports[`parseAst 自定义组件按需导入 1`] = `
import withWeapp, { cacheOptions } from "@tarojs/with-weapp";
import { Block } from "@tarojs/components";
import React from "react";
import Taro from "@tarojs/taro";
import ComponentB from '../components/b';
import ComponentA from '../components/a';
cacheOptions.setOptionsToCache({});
@withWeapp(cacheOptions.getOptionsFromCache())
class _C extends React.Component {
  render() {
    return <Block><ComponentA name="a"></ComponentA><ComponentB name="b"></ComponentB>{
        /*  <component-c name="c" /> */
      }</Block>;
  }
}
export default _C;
`;

exports[`parseAst 转换后模块名重名时，为导入的组件模块名添加Component后缀，以示区分 1`] = `
import withWeapp, { cacheOptions } from "@tarojs/with-weapp";
import { Block } from "@tarojs/components";
import React from "react";
import Taro from "@tarojs/taro";
import Jstest from '../../components/test.js';
import { Mymodule, Mymodules } from '../../components/test.js';
import MymodulesComponent from '../../components/my-component';
import MymoduleComponent from '../../components/my-component';
import JstestComponent from '../../components/my-component';
cacheOptions.setOptionsToCache({});
@withWeapp(cacheOptions.getOptionsFromCache())
class _C extends React.Component {
  render() {
    return <Block><JstestComponent></JstestComponent><MymoduleComponent></MymoduleComponent><MymodulesComponent></MymodulesComponent></Block>;
  }
}
export default _C;
`;
